### DISCUSSION TITLE
Site not updated, clear cache manually

### DISCUSSION THREAD

**Discussion by ghost** [datenstrom/community#48] [Report a bug] [answered]
The page will not get updated automatically on certain scenarios:
- move folders in the content folder (e.g. change name from feed to 3-feed)
- add new folders to content folder
- editing a blog entry (with or without images) will not update the front page (Add a false image link, then add a correct one. The / and /blog-entry will not get updated)

I think, that the root page is some kind of broken concerning updates. Only clearing my browser cache helps. This has to be fixed like as soon as possible. :+1:

**Comment by markseuffert** [comment]
About the issue with emptying cache. Yes, this happens when you make structural changes on your website... such as snippets, templates and folders. When this happens reload page or empty the browser cache. Sorry about that. Let's see if we can find a solution.

About the issue with editing a blog entry, I was not able to reproduce it.

Thanks a lot for reporting, any additional information is appreciated!

**Comment by nibreh** [comment]
HI

By the way, I found this trick in stackoverflow :

header("Expires: Tue, 01 Jan 2000 00:00:00 GMT");
header("Last-Modified: " . gmdate("D, d M Y H:i:s") . " GMT");
header("Cache-Control: no-store, no-cache, must-revalidate, max-age=0");
header("Cache-Control: post-check=0, pre-check=0", false);
header("Pragma: no-cache");

Put it in _yellow.php_ before _require_once("system/core/core.php");_

_This doesn't clear the user's browser cache. It does tell the browser to not cache the site._
http://stackoverflow.com/questions/1037249/how-to-clear-browser-cache-with-php

Don't know if it's a good trick or not, but it's useful when testing yellowCMS

**Comment by markseuffert** [comment]
I didn't know this was such a big problem, that people were using workarounds.

**Comment by nibreh** [comment]
Well, it's not really a problem for me, because I don't save cache browser (delete automatically when close browser). But most of people don't clear their cache even after years ;)

**Comment by markseuffert** [comment]
Good to know. I was not aware it was a problem. :smiley: 

Here's what I understand so far. [Caching is a good thing](https://www.mnot.net/cache_docs/), it reduced latency and traffic. But sometimes the browser shows an old page from the browser cache. This happens for two reasons: The browser doesn't request a new page or the CMS says that there is no new page. First case is usually not a problem for small websites, you simply click reload. Second case is the bigger problem. For example after changing content folders, the CMS doesn't tell the browser that the navigation has changed. Then you have to clean the browser cache. 

Let's see how we can improve it. Ideas, suggestions?

**Comment by ghost** [comment]
I do think that @nibreh's patch should be included by default as long as this issue isn't fixed correctly using caches. I know many people who don't press F5 to reload. They just stare at the page and see no changes at all. As the maintainer of a website, I really want people to see all UPDATED contents without the help of clearing the cache or reloading the whole site. They might do that though, when they think the page should have some updated information, so traffic is not the main problem for most of the visitors. It is about comfort.

**Comment by markseuffert** [comment]
The latest version should fix caching problems. Let me know how it works, thanks!

**Comment by ghost** [comment]
Seems to be fixed. I am closing it.

**Comment by ghost** [comment]
By the way. This is **NOT** fixed. Seems still to be a problem, when I upload a new .txt file (e.g. new blog post). After pressing F5 it updates the site. Should be working without having to press F5.

**Comment by markseuffert** [comment]
Does this happen on your blog start page? I think what's happening is that the browser doesn't request a new page. The visitor sees the old page until someone decides to get a new page from the CMS. Let's try a cache busting/expiration technique. Add the following line to your blogarticles template:

```
<?php $yellow->page->header("Cache-Control: max-age=60") ?>
```

This should force the browser to request a new page. Let me know how it works.

**Comment by ghost** [comment]
Not working. Added it on top of the blogarticles template file.

**Comment by markseuffert** [comment]
Let's facetime/skype?

**Comment by ghost** [comment]
I don't have those things and currently no time. Maybe in 1h. Do you have jabber?

**Comment by markseuffert** [comment]
It was just a suggestion, sometimes things are faster to fix on the phone.  :smile: 

Sorry don't have Jabber. Can you try a smaller timeout value, e.g. 10 seconds? From my side it looks good, the Cache-Control header is in the HTTP response. I tested your blog, could also confirm that the browser throws away the cached blog page after the timeout and tries to get a new page from the server. 

Here some timeouts that news sites use:

60 http://www.theguardian.com/uk
32 http://www.heise.de/newsticker/
0 http://www.sydsvenskan.se/sverige/

0 means no cache, feel free to experiment.

**Comment by ghost** [comment]
Ok, I think it is working. Should this be default on blogarticles template? :>

**Comment by markseuffert** [comment, marked as answer]
Yes, templates are updated. Should be working now.
